`crypto/evp`: compensate for providers not adding error queue entries for keymgmt, sigver, and asymcipher

DDvO opened the pull request:
Unfortunately, crypto providers often do not add any error queue entries in case of failures.
This makes it for developers and users needlessly hard to narrow down where and why a crypto error occurred.

This PR compensates for this issue by adding error queue entries for keymgmt, sigver, and asymcipher operation failures,
giving the key type, name of the provider function called, and the operation description set by the provider, if any.

This PR also adds error queue entries in case an expected operation function is not implemented by a provider.
This can be seen as a partial fix for #22940.

On this occasion, in the error case of `ossl_store_handle_load_result()` in `store_result.c`,
add to the error queue which provider failed to load a credential and a hint on using default provider if it is not available.

paulidale commented on crypto/evp/m_sigver.c:
```diff
@@ -633,8 +683,19 @@ int EVP_DigestVerifyFinal(EVP_MD_CTX *ctx, const unsigned char *sig,
+    signature = pctx->op.sig.signature;
+    desc = signature->description != NULL ? signature->description
+        : "no signature description by provider";
+    if (signature->digest_verify_final == NULL) {
```
These added NULL checks aren't required.  `signature.c`  checks that if the first of the calls is defined then the rest are too:

https://github.com/openssl/openssl/blob/8d2e4d6d8c927f05948e048fcbf62982feaf11b4/crypto/evp/signature.c#L364-L372

Consequently, it's only necessary to check if the init function is not NULL.

DDvO replied:
Thanks for pointing out these checks - very Interesting, also because they (in conjunction with the documentation at doc/man7/provider-signature.pod) meanwhile implement at least major part of the guidance reported missing at #22940.

Two points to consider:
1. Since functions like `EVP_DigestVerifyUpdate()` are API functions, is it always guaranteed (via the `EVP_MD_CTX *ctx` parameter) that `evp_signature_from_algorithm()`, the internal function that performs a lot of such checks, has been called before with positive outcome?
2. The logic implemented by `evp_signature_from_algorithm()` appears incomplete.
In particular, the presence of `digest_verify_update` and `digest_verify_final` is checked only if `digest_verify == NULL`,
which apparently implies that they can still be called if the provider only implements `digest_init` and `digest_verify`.
Moreover, in a few lines before,
```
    if (valid
        && (signature->verify != NULL
            || signature->verify_message_update != NULL
            || signature->verify_message_final != NULL)
        && signature->verify_init == NULL
        && signature->verify_message_init == NULL)
```
likely should be replaced by
```
    if (valid
        && (signature->verify != NULL
            signature->verify_init == NULL)
        || ((signature->verify_message_update != NULL
             || signature->verify_message_final != NULL)
            && signature->verify_message_init == NULL)
 ```

DDvO replied:
I think when `evp_signature_from_algorithm()` is called, it is not yet known whether the one-shot functions or the ones involving "update" and "final" will be used, and therefore believe that this function cannot do a sufficient check on the availability of, e.g., `digest_verify_update` and `digest_verify_final`.

paulidale reviewed (commented):
Nice work.

Not sure we want the over-paranoia function pointer NULL checks since they're all known to be safe by the time they are reached.

t8m commented on crypto/evp/asymcipher.c:
```diff
@@ -159,18 +160,22 @@ static int evp_pkey_asym_cipher_init(EVP_PKEY_CTX *ctx, int operation,
     }
 
+    desc = cipher->description != NULL ? cipher->description
+        : "no asym cipher description by provider";
```
This indentation looks bad to me. Also the text is IMO too bulky - I would at least remove the _by provider_ part.

DDvO replied:
The indentation is fine according the the Emacs C mode and the `check-format.pl` tool, 
but anyway, after chopping the " by provider" part, the RHS is less wide and fits into the 100 char limit.

t8m commented on crypto/evp/asymcipher.c:
```diff
@@ -159,18 +160,22 @@ static int evp_pkey_asym_cipher_init(EVP_PKEY_CTX *ctx, int operation,
     case EVP_PKEY_OP_ENCRYPT:
         if (cipher->encrypt_init == NULL) {
-            ERR_raise(ERR_LIB_EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
+            ERR_raise_data(ERR_LIB_EVP, EVP_R_PROVIDER_ASYM_CIPHER_MISSING,
```
Why do you change the reason code? IMO the original one was better.  If you want you could put `asym_cipher_encrypt_init` into the text below to make it clear what operation is not provided.

DDvO replied:
I've removed all the tentative new `EVP_R_PROVIDER_*_MISSING` reason codes
and now use in all such cases `EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE`.

DDvO commented:
BTW, looks like there are more such bugs not fixed by this PR, namely for "sign_message" and "verify_message".

t8m commented:
> @t8m I'm pretty sure that part of this PR actually fixes bugs, namely the potential to call NULL implementations of "update" and "final" functions in case just the respective "init" function is present.

This is still IMO not a bug in a sense this should be backported to stable branches. Such provider is dysfunctional and should not be used.

DDvO commented:
Also added a commit for: `provider-signature.pod`: 
add missing doc of `OSSL_FUNC_signature_query_key_types()` and fix doc of return types

paulidale commented:
ML-DSA and SLH-DSA do not include the update and final methods.

IMO this should go back to 3.5 so that our FIPS and default providers do not fall under the _dysfunctional and should not be used_ classification.

DDvO commented:
Added a commit: check in `crypto/evp/signature.c` for consistent presence of 'update' and 'final' functions
since, as [mentioned above](https://github.com/openssl/openssl/pull/27367#discussion_r2041704709), for each sign or verify operation, the "final" function must be available if and only if the respective "update" function is present.

DDvO commented:
Rebased to fix trivial merge conflict;
on this occasion also added two nit fixes suggested in https://github.com/openssl/openssl/pull/27372#discussion_r2046339544 and https://github.com/openssl/openssl/pull/27372#pullrequestreview-2771562790.

DDvO commented:
Ping @openssl/committers for reviews/approvals

vdukhovni commented on crypto/store/store_result.c:
```diff
@@ -153,8 +153,17 @@ int ossl_store_handle_load_result(const OSSL_PARAM params[], void *arg)
 
-    if (*v == NULL)
-        ERR_raise(ERR_LIB_OSSL_STORE, ERR_R_UNSUPPORTED);
+    if (*v == NULL) {
+        const char *prov_name = NULL, *hint = "";
+
+        if (provider != NULL)
+            prov_name = OSSL_PROVIDER_get0_name(provider);
+        if (prov_name == NULL)
+            prov_name = "(none)";
+        if (!OSSL_PROVIDER_available(libctx, "default"))
+            hint = ":maybe need to load the default provider?";
+        ERR_raise_data(ERR_LIB_OSSL_STORE, ERR_R_UNSUPPORTED, "provider=%s%s", prov_name, hint);
+    }
```
Again, I'd suggest a simple rewrite here, where absent a provider, instead if "(none)", suppress the entire "provider=" part, and if the default provider is loaded (no hint), perhaps suppress the entire detail string.

DDvO replied:
done

vdukhovni commented on doc/man7/provider-signature.pod:
```diff
@@ -201,10 +200,22 @@ set of "signature" functions, i.e. at least one of:
-but if one of them is present then the other one must also be present. The same
-applies to OSSL_FUNC_signature_get_ctx_params and OSSL_FUNC_signature_gettable_ctx_params, as
-well as the "md_params" functions. The OSSL_FUNC_signature_dupctx function is optional.
+OSSL_FUNC_signature_set_ctx_params and OSSL_FUNC_signature_settable_ctx_params
+function are optional,
```
```suggestion
The OSSL_FUNC_signature_set_ctx_params() and OSSL_FUNC_signature_settable_ctx_params()
functions are optional,
```

DDvO replied:
Done, also for all doc improvement suggestions below.

vdukhovni commented on doc/man7/provider-signature.pod:
```diff
@@ -201,10 +200,22 @@ set of "signature" functions, i.e. at least one of:
-well as the "md_params" functions. The OSSL_FUNC_signature_dupctx function is optional.
+OSSL_FUNC_signature_set_ctx_params and OSSL_FUNC_signature_settable_ctx_params
+function are optional,
+but if one of them is present then the other one must also be present.
+The same applies to the
```
```suggestion
but if one of them is provided then the other one must also be provided.
The same applies to the
```

vdukhovni commented on doc/man7/provider-signature.pod:
```diff
@@ -201,10 +200,22 @@ set of "signature" functions, i.e. at least one of:
+as well as the "md_params" functions.
+
+The OSSL_FUNC_signature_dupctx function is optional.
+So far it is not used by OpenSSL.
```
```suggestion
It is not yet used by OpenSSL.
```

vdukhovni commented on doc/man7/provider-signature.pod:
```diff
@@ -201,10 +200,22 @@ set of "signature" functions, i.e. at least one of:
+function are optional,
+but if one of them is present then the other one must also be present.
+The same applies to the
+OSSL_FUNC_signature_get_ctx_params and OSSL_FUNC_signature_gettable_ctx_params functions,
```
```suggestion
OSSL_FUNC_signature_get_ctx_params() and OSSL_FUNC_signature_gettable_ctx_params() functions,
```

vdukhovni commented on doc/man7/provider-signature.pod:
```diff
@@ -201,10 +200,22 @@ set of "signature" functions, i.e. at least one of:
+OSSL_FUNC_signature_get_ctx_params and OSSL_FUNC_signature_gettable_ctx_params functions,
+as well as the "md_params" functions.
+
+The OSSL_FUNC_signature_dupctx function is optional.
```
```suggestion
The OSSL_FUNC_signature_dupctx() function is optional.
```

vdukhovni commented on doc/man7/provider-signature.pod:
```diff
@@ -201,10 +200,22 @@ set of "signature" functions, i.e. at least one of:
+The OSSL_FUNC_signature_dupctx function is optional.
+So far it is not used by OpenSSL.
+
+The OSSL_FUNC_signature_query_key_types function is optional.
```
And other similar forms. Perhaps literal function names that are not references to other docs (`L<...>`) should perhaps be enclosed in `B<...>`.
```suggestion
The OSSL_FUNC_signature_query_key_types() function is optional.
```

DDvO commented on crypto/evp/asymcipher.c:
```diff
@@ -159,18 +160,21 @@ static int evp_pkey_asym_cipher_init(EVP_PKEY_CTX *ctx, int operation,
         goto err;
     }
 
+    desc = cipher->description != NULL ? cipher->description : "no asym cipher description";
```
The string added in case `cipher->description == NULL ` is not content-free.
It deliberately adds the hint that error thrown is about an asymmetric cipher and that the provider did not give a description for it.
Likewise for similar cases below.

vdukhovni replied:
Well, if you want to always indicate that an asymmetric cipher operation failed, say so in the fixed format template. Lack of a description string is NOT useful text to report to the user.

t8m replied:
I agree with @vdukhovni this should be in the template.

DDvO replied:
So, you mean, I should replace the respective `EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE` by e.g,
```
EVP_R_PROVIDER_ASYM_CIPHER_NOT_SUPPORTED
EVP_R_PROVIDER_KEYMGMT_NOT_SUPPORTED
EVP_R_PROVIDER_SIGNATURE_NOT_SUPPORTED
```
in analogy to the new
```
EVP_R_PROVIDER_ASYM_CIPHER_FAILURE:232:provider asym cipher failure
EVP_R_PROVIDER_KEYMGMT_FAILURE:233:provider keymgmt failure
EVP_R_PROVIDER_SIGNATURE_FAILURE:234:provider signature failure
```
?

DDvO replied:
When the description is not given by the provider, I realized that the extra `desc` variable is not really needed,
because error calls could simply look like this:
```
ERR_raise_data(ERR_LIB_EVP, EVP_R_PROVIDER_ASYM_CIPHER_NOT_SUPPORTED,
                           "%s encrypt_init:desc=%s", cipher->type_name, cipher->description);
```
where in case `cipher->description == NULL`, conveniently `<null>` is inserted.
Okay?

t8m replied:
Hmm... we could depend on NULL being converted to `<null>` because it is printed by our own BIO_vsnprintf() but there were discussions of using snprintf() eventually which AFAIK might not handle it gracefully?

I'd keep NULL -> "" conversion.

DDvO replied:
Done, also for all similar cases below.

DDvO commented on crypto/evp/signature.c:
```diff
@@ -315,64 +324,129 @@ static void *evp_signature_from_algorithm(int name_id,
+    if ((signature->sign_message_update == NULL) !=
+        (signature->sign_message_final == NULL)) {
+        ERR_raise_data(ERR_LIB_EVP, EVP_R_INVALID_PROVIDER_FUNCTIONS,
+                       "inconsistent presence of %s message signing update and final:%s",
```
Done,  also for all similar cases below.

DDvO commented:
Thanks @roman for your nice feedback!

Regarding backporting to 3.5, this has already been indicated by @paulidale (setting the `branch: 3.5` label),
and nobody objected during also the subsequent reviews.
So when this is ready to merge, looks like this can be cherry-picked also to 3.5.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been  updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

t8m commented:
Merged to the master and 3.5 branches. Thank you for your contribution.
